### PR TITLE
Windows: Fix encoding of `Dir.home` and prefer `USERPROFILE` over `HOMEPATH`

### PR DESCRIPTION
**Pull request by larskanis**
`Dir.home` returns an UTF-8 string since ruby-3.0, but the actual encoding of the bytes was `CP_ACP` or `CP_OEMCP`. That led to invalid bytes when calling `Dir.home` with an Unicode username.

`HOMEPATH` is set to `\WINDOWS\system32` when running per `runas` session. This directory is not writable by ordinary users, leading to errors with many ruby tools. Also config files in the home directory are not recognized.

Still keeping `HOME` at first, which is not used by native Windows, but by ruby specs and by MSYS2 environment.

This might be backported to ruby-3.[01].

### CODE REVIEW

File: win32/win32.c [hunk 1]
```diff
@@ -579,13 +579,13 @@ rb_w32_home_dir(void)
...
+      case ENV_USERPROFILE:
+        GetEnvironmentVariableW(L"USERPROFILE", buffer, buffer_len);
+        break;
       case ENV_DRIVEPATH:
         len = GetEnvironmentVariableW(L"HOMEDRIVE", buffer, buffer_len);
         GetEnvironmentVariableW(L"HOMEPATH", buffer + len, buffer_len - len);
         break;
-      case ENV_USERPROFILE:
-        GetEnvironmentVariableW(L"USERPROFILE", buffer, buffer_len);
-        break;
```
Comment (nobu) [inline win32/win32.c:588]: Seems no effect.

Comment (larskanis) [reply win32/win32.c:588]: Sure, but it's confusing to have the cases in a different order than the if's above.

**Comment by larskanis** [comment]
May I open two corresponding bug reports on bugs.ruby-lang.org for completeness?

**Comment by larskanis** [comment]
I opened two bug reports on bugs.ruby-lang.org regarding:
* [wrong encoding](https://bugs.ruby-lang.org/issues/19243)
* [prefer USERPROFILE](https://bugs.ruby-lang.org/issues/19244)

File: win32/win32.c (hunk 1 above)
Comment (nobu) [reply win32/win32.c:588]: Then `enum home_type` should be reordered too?

Comment (larskanis) [reply win32/win32.c:588]: You're right - I changed the enum as well.

**Comment by nobu** [comment]
Could you split PRs for two issues?

**Comment by larskanis** [comment]
OK, the PR is split into #6991 and #6992 and the specs are more separated now.

**Pull request closed without merging**